### PR TITLE
CDAP-12151 introduce AlertEmitter and AlertPublisher APIs

### PR DESCRIPTION
**Pull request by albertshau**
Add a new plugin type for AlertPublisher and add a method to all
emitters to emit alerts.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/AlertEmitter.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.api;
+
+import java.util.Map;
+
+/**
+ * Emits alerts.
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/AlertEmitter.java:22]: nit: `Emit alerts.`

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformingEmitter.java [hunk 2]
```diff
@@ -66,4 +72,5 @@ public void emitError(InvalidEntry<EMIT> value) {
   public void reset() {
     emitter.reset();
   }
+
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformingEmitter.java]: nit: extra line

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/NoErrorEmitter.java [hunk 3]
```diff
@@ -47,11 +48,17 @@ public void emitError(InvalidEntry<T> value) {
...
+  }
+
   public Collection<T> getEntries() {
     return entryList;
   }
 
   public void reset() {
     entryList.clear();
   }
+
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/NoErrorEmitter.java]: nit: extra line

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultEmitter.java [hunk 4]
```diff
@@ -47,16 +52,27 @@ public void emitError(InvalidEntry<T> value) {
...
+  public Collection<Map<String, String>> getAlerts() {
+    return Collections.unmodifiableList(alerts);
   }
 
   public void reset() {
     entryList.clear();
     errorList.clear();
+    alerts.clear();
   }
+
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultEmitter.java]: nit: extra line

**Review by sagarkapare — APPROVED** [review]
We updated interfaces, however seems like the change would be backward compatible. worth verifying otherwise minor nits and LGTM 👍

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT5961

**Comment by albertshau** [comment]
verified it's binary compatible. expanded superinterface set, and this interface is not meant to be implemented by clients, so adding a method is ok too.

**Pull request merged**